### PR TITLE
feat: define API and add reference implementation for KMS server

### PR DESCRIPTION
**Pull request by Unix4ever**
API has `Seal` and `Unseal` methods, which will be called from Talos key handler implementation for KMS based disk encryption.

### CODE REVIEW

File: hack/release.toml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# commit to be tagged for the new release
+commit = "HEAD"
+
+project_name = "siderolink"
```
Comment (frezbo) [inline hack/release.toml:4]:
```suggestion
project_name = "kms-client"
```

File: hack/release.toml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+github_repo = "talos-systems/siderolink"
```
Comment (frezbo) [inline hack/release.toml:5]: this too

File: pkg/server/server.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+// NewServer initializes new server.
+func NewServer(keyHandler func(nodeUUID string) ([]byte, error)) *Server {
+	return &Server{
+		getKey: keyHandler,
+	}
+}
+
+// Seal encrypts the incoming data.
+func (srv *Server) Seal(ctx context.Context, req *kms.Request) (*kms.Response, error) {
+	time.Sleep(time.Second)
```
Comment (frezbo) [inline pkg/server/server.go:35]: needed?

**Review by frezbo — APPROVED** [review]

File: pkg/server/server.go (hunk 3 above)
Comment (Unix4ever) [reply pkg/server/server.go:35]: anti brute-force defence :D

**Comment by Unix4ever** [comment]
I wonder if we need to do something in drone as the build doesn't start :thinking: 
I don't see the project in the list to activate it too.

**Comment by frezbo** [comment]
> I wonder if we need to do something in drone as the build doesn't start 🤔 I don't see the project in the list to activate it too.

https://ci.dev.talos-systems.io/siderolabs/kms-client/settings

this maybe?

**Comment by Unix4ever** [comment]
Ok I don't see it as I'm not an admin.

File: pkg/server/server.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	}
+
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, err
+	}
+
+	nonce := make([]byte, 12)
+	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
+		panic(err.Error())
```
Comment (smira) [inline pkg/server/server.go:57]: panic seems to be a bit of overkill here ;)

File: pkg/server/server.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, err
+	}
+
+	decrypted, err := aesgcm.Open(nil, req.Data[:12], req.Data[12:], nil)
```
Comment (smira) [inline pkg/server/server.go:89]: I wonder if the method should return some garbage (random data) if the `Open` fails to prevent brute-forcing.

Comment (Unix4ever) [reply pkg/server/server.go:89]: or just return back encrypted string?

File: api/kms/kms.proto [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+service KMSService {
+  rpc Seal(Request) returns (Response);
+  rpc Unseal(Request) returns (Response);
+}
+
+message Request {
+  // Node UUID as string.
+  string node_uuid = 1;
+  // Data to encrypt or decrypt.
+  bytes data = 2;
```
Comment (smira) [inline api/kms/kms.proto:17]: we should probably limit the `data` length here, more for consistency reasons? 

Seal/Unseal have a bit different requirements, but still

Comment (Unix4ever) [reply api/kms/kms.proto:17]: What limit do you think will make sense?

Comment (smira) [reply api/kms/kms.proto:17]: this is another good topic, I was going to ask this question in Talos PR. 

e.g. can we do 32-bytes random encryption key, that should be more than enough.

so on Seal, we can say `len(data) == 32`, and Unseal should be `32 + 12` always

This should match TPM key size

File: pkg/server/server.go (hunk 5 above)
Comment (smira) [reply pkg/server/server.go:89]: if we check the length, we can cut away `nonce`, and return random expected 32 bytes for example

File: api/kms/kms.proto (hunk 6 above)
Comment (Unix4ever) [reply api/kms/kms.proto:17]: Looks like TPM key size is in range 32 - 128 byte

Comment (smira) [reply api/kms/kms.proto:17]: the actual encryption key (unsealed one) is 32 bytes: https://github.com/systemd/systemd/blob/06afda6b38d5d730fca3c65449096425933272bc/src/shared/tpm2-util.c#L2918-L2937

Comment (Unix4ever) [reply api/kms/kms.proto:17]: > and Unseal should be 32 + 12 always

I don't think it's true: encrypted data is 48 bytes + nonce it gets to 60.

Comment (Unix4ever) [reply api/kms/kms.proto:17]: that's because AES has fixed block of 16 bytes. So I guess I can just make it 60

Comment (smira) [reply api/kms/kms.proto:17]: yep, sorry, sealed is longer due to auth, thanks

File: hack/release.toml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+# commit to be tagged for the new release
+commit = "HEAD"
+
+project_name = "kms-client"
+github_repo = "talos-systems/kms-client"
```
Comment (frezbo) [inline hack/release.toml:5]:
```suggestion
github_repo = "siderolabs/kms-client"
```

File: api/kms/kms.proto (hunk 6 above)
Comment (Unix4ever) [reply api/kms/kms.proto:17]: fixed

File: pkg/server/server.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, err
+	}
+
+	nonce := make([]byte, 12)
```
Comment (smira) [inline pkg/server/server.go:58]: we should probably use `aesgcm.NonceSize()` instead of hardcoded value here

File: pkg/server/server.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
+		return nil, err
+	}
+
+	if len(req.Data) != 32 {
```
Comment (smira) [inline pkg/server/server.go:63]: 32 should be a constant I guess

File: pkg/server/server.go [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+	return &kms.Response{
+		Data: append(nonce, encrypted...), //nolint:makezero
+	}, nil
+}
+
+// Unseal decrypts the incoming data.
+func (srv *Server) Unseal(_ context.Context, req *kms.Request) (*kms.Response, error) {
+	time.Sleep(time.Second)
+
+	if len(req.Data) != 60 {
```
Comment (smira) [inline pkg/server/server.go:78]: should not use magic value here, `aesgcm.` has some methods to know the length

File: pkg/server/server.go [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+	}
+
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := &kms.Response{}
+
+	decrypted, err := aesgcm.Open(nil, req.Data[:12], req.Data[12:], nil)
```
Comment (smira) [inline pkg/server/server.go:102]: same about nonce size here (not using magic 12)

File: pkg/server/server.go [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+		return resp, nil
+	}
+
+	resp.Data = decrypted
+
+	return resp, nil
+}
+
+// GetRandomAESKey generates random AES256 key.
+func GetRandomAESKey() ([]byte, error) {
```
Comment (smira) [inline pkg/server/server.go:119]: should be private?

**Review by smira — APPROVED** [review]

**Pull request merged**